osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -0,0 +1,168 @@
+            return current.getSecret().equals(secret.getSecret());
+        } catch (final NoSuchElementException nsee) {
+            //there is no current secret
+            return false;
+        }
+    }
+
+    private DbEncryptedSecret createEncrypted(final String deviceIdentification, final TypedSecret typedSecret,
+            final DbEncryptionKeyReference keyReference) {
+        final String secretString = typedSecret.getSecret();
+        final byte[] secretBytes = HexUtils.fromHexString(secretString);
+        final Secret secret = new Secret(secretBytes);
+        try {
+            final EncryptedSecret encryptedSecret = this.encryptionDelegate.encrypt(
+                    keyReference.getEncryptionProviderType(), secret, keyReference.getReference());
+            final DbEncryptedSecret dbEncryptedSecret = new DbEncryptedSecret();
+            dbEncryptedSecret.setDeviceIdentification(deviceIdentification);
+            dbEncryptedSecret.setEncodedSecret(HexUtils.toHexString(encryptedSecret.getSecret()));
+            dbEncryptedSecret.setSecretType(typedSecret.getSecretType());
+            dbEncryptedSecret.setEncryptionKeyReference(keyReference);
+            dbEncryptedSecret.setCreationTime(new Date());
+            return dbEncryptedSecret;
+        } catch (final Exception exc) {
+            throw new IllegalStateException("Could not create encrypted secret", exc);
+        }
+    }
+
+    public List<TypedSecret> retrieveSecrets(final String deviceIdentification, final List<SecretType> secretTypes)
+            throws Exception {
+        try {
+            //@formatter:off
+            return secretTypes.stream()
+                    .map(secretType -> this.retrieveSecret(deviceIdentification,secretType))
+                    .collect(Collectors.toList());
+            //@formatter:on
+        } catch (final Exception exc) {
+            throw new IllegalStateException(
+                    String.format("Something went wrong retrieving secrets for device %s", deviceIdentification), exc);
+        }
+    }
+
+    @java.lang.SuppressWarnings("squid:S3655")
+    public TypedSecret retrieveSecret(final String deviceIdentification, final SecretType secretType) {
+        final Date now = new Date();
+        final Long secretId = this.secretRepository.findIdOfValidMostRecent(deviceIdentification, secretType.name(),
+                now);
+        if (secretId == null) {
+            throw new NoSuchElementException("No secret found with a valid key");
+        }
+        return this.getTypedSecret(this.secretRepository.findById(secretId).get());
+    }
+
+    private TypedSecret getTypedSecret(final DbEncryptedSecret dbEncryptedSecret) {
+        if (dbEncryptedSecret != null) {
+            final DbEncryptionKeyReference keyReference = dbEncryptedSecret.getEncryptionKeyReference();
+            if (keyReference == null) {
```

Comment:
and another null check?

consider validating the DbEncryptedSecret  first before calling this method